apps/ledger-live-desktop/src/renderer/components/CustomImage/ImageGrayscalePreview.tsx

```diff
@@ -0,0 +1,219 @@
+import React, { useCallback, useEffect, useRef, useState } from "react";
+import { Flex } from "@ledgerhq/react-ui";
+import { createCanvas } from "./imageUtils";
+import { ImageBase64Data, ImageDimensions } from "./types";
+import { targetDimensions } from "./shared";
+import ContrastChoice from "./ContrastChoice";
+import { ImageProcessingError } from "@ledgerhq/live-common/customImage/errors";
+
+export type ProcessorPreviewResult = ImageBase64Data & ImageDimensions;
+export type ProcessorRawResult = { hexData: string } & ImageDimensions;
+
+export type ProcessorResult = {
+  previewResult: ProcessorPreviewResult;
+  rawResult: ProcessorRawResult;
+};
+
+const contrasts = [
+  { val: 1, color: "neutral.c70" },
+  { val: 1.5, color: "neutral.c50" },
+  { val: 2, color: "neutral.c40" },
+  { val: 3, color: "neutral.c30" },
+];
+
+function clamp(val: number, min: number, max: number) {
+  return Math.min(max, Math.max(min, val));
+}
+
+function contrastRGB(rgbVal: number, contrastVal: number) {
+  return (rgbVal - 128) * contrastVal + 128;
+}
+
+// simutaneously apply grayscale and contrast to the image
+function applyFilter(
+  imageData: ImageData,
+  contrastAmount: number,
+  dither = true,
+): { imageDataResult: Uint8ClampedArray; hexRawResult: string } {
+  let hexRawResult = "";
+  const filteredImageData = [];
+
+  const data = imageData.data;
+
+  const numLevelsOfGray = 16;
+  const rgbStep = 255 / (numLevelsOfGray - 1);
+
+  const { width, height } = imageData;
+
+  const pixels256: number[][] = Array.from(Array(height), () => Array(width));
+  const pixels16: number[][] = Array.from(Array(height), () => Array(width));
+
+  for (let pxIndex = 0; pxIndex < data.length / 4; pxIndex += 1) {
+    const x = pxIndex % width;
+    const y = (pxIndex - x) / width;
+
+    const [redIndex, greenIndex, blueIndex] = [4 * pxIndex, 4 * pxIndex + 1, 4 * pxIndex + 2];
+    const gray256 = 0.299 * data[redIndex] + 0.587 * data[greenIndex] + 0.114 * data[blueIndex];
+
+    /** gray rgb value after applying the contrast */
+    pixels256[y][x] = clamp(contrastRGB(gray256, contrastAmount), 0, 255);
+  }
+
+  for (let y = 0; y < height; y++) {
+    for (let x = 0; x < width; x++) {
+      const oldpixel = pixels256[y][x];
+      const posterizedGray16 = Math.floor(oldpixel / rgbStep);
+      const posterizedGray256 = posterizedGray16 * rgbStep;
+      /**
+       * Floyd-Steinberg dithering
```

Comment:
❤️👨‍🎨

Reply:
🥂 a toast to the old masters of image processing